[Checkup][Ability] Add Steadfast Ability Checkup

flx-sta opened the pull request:
## What are the changes the user will see?

- `<NAME> has flinched!` will no longer appear AFTER seeing the speed-boost message of the Steadfast ability
- Steadfast will no longer boost the owner speed if flinch would occur after it already acted

## Why am I making these changes?

Closes #721 
Closes #529 

## What are the changes from a developer perspective?

- Add test-coverage for Steadfast
- Fix `FlinchStatStageChangeAbAttr` being applied upon any `FlinchedTag` add
    - Instead it's now only applied when the `FlinchedTag` is lapsing via `PRE_MOVE`
    - This fixes #529 and also prevents the "Speed up" message appearing BEFORE `flinched` message appeared
- Add custom matchers:
    - `toHaveFlinched`
   - `toHaveStatStage`

## Screenshots/Videos

#### Before

<details><summary>[🎥 Show Video]</summary>
<p>

https://github.com/user-attachments/assets/895dcb9c-df0f-4ecb-aeeb-c475152ecd88

</p>
</details> 

### After

<details><summary>[🎥 Show Video]</summary>
<p>

https://github.com/user-attachments/assets/4cce4193-7f03-4008-8257-67acdd47eef8

</p>
</details> 

## How to test the changes?

Run the automated tests

### Overrides

```ts
const overrides = {
  ENEMY_MOVESET_OVERRIDE: MoveId.FAKE_OUT,
  STATUS_OVERRIDE: StatusEffect.PARALYSIS,
  ABILITY_OVERRIDE: AbilityId.STEADFAST,
} satisfies Partial<InstanceType<typeof DefaultOverrides>>;
```

`STATUS_OVERRIDE: StatusEffect.PARALYSIS` So our Pokemon is (almost) guaranteed to be slower!

## Checklist

- [x] Otherwise: **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I tested the changes manually?
- [x] Are all unit tests still passing? (`npm run test:silent`)
  - [x] Have I created new automated tests (`npm run test:create`) or updated existing tests related to the PR's changes?

flx-sta committed:
Reduce `loaded locale` noise in tests

- it can still be printed by setting "VITE_I18N_DEBUG=1"

flx-sta committed:
Add 2 matchers

- to have flinched matcher
- to have stat stage matcher

flx-sta committed:
Add tsdoc to FlinchStatStageChangeAbAttr

flx-sta committed:
fix [Bug] Flinching Can Occur After a Pokemon Has Moved #529

- Make FlinchStatStageChangeAbAttr only apply if the pokemon hasn't acted yet

flx-sta committed:
Add Steadfast tests

flx-sta committed:
Fix: FlinchStatStageChangeAbAttr application lifecycle

the tag was originally applied immediately after flinching, which cause the attribute to also be triggered when the flinching occured AFTER the owner already acted. Now it's only applied in lapse and only if the lapse type ise "pre-move".

flx-sta committed:
update Steadfast tests

Tempo-anon commented on test/abilities/steadfast.test.ts:
```diff
@@ -0,0 +1,87 @@
+    expect(playerPkm).toHaveStatStage(Stat.SPD, +1);
+  });
+
+  it(`should NOT boost SPD after NOT flinching`, async () => {
```
The wording on this is a bit awkward. Maybe `should NOT boost SPD when pokemon does not flinch`
```suggestion
  it(`should NOT boost SPD when Pokemon does NOT flinch`, async () => {
  ```

Tempo-anon commented on test/abilities/steadfast.test.ts:
```diff
@@ -0,0 +1,87 @@
+    expect(playerPkm).toHaveStatStage(Stat.SPD, 0);
+  });
+
+  it(`should NOT boost SPD if flinchinging occured after owner acted`, async () => {
```
typo: flinching

```suggestion
  it(`should NOT boost SPD if flinching occured after owner acted`, async () => {
```

Tempo-anon commented on test/abilities/steadfast.test.ts:
```diff
@@ -0,0 +1,87 @@
+    const playerPkm = field.getPlayerPokemon();
+
+    game.setTurnOrder([BattlerIndex.PLAYER, BattlerIndex.ENEMY]);
+    move.use(MoveId.SPLASH);
+    await move.selectEnemyMove(MoveId.FAKE_OUT);
```
Does this work like it's supposed to? To force a move with higher priority to go after. I'd think calling a move with flinching like bite/air slash makes a bit more sense

flx-sta replied:
It does indeed work.
I'm using `game.setTurnOrder([BattlerIndex.PLAYER, BattlerIndex.ENEMY]);` to force the order of moves as desired.

## Test Output

<details><summary>[📄 Show Output]</summary>
<p>

```prolog
Start Phase EnemyCommandPhase 
Start Phase TurnStartPhase 
Start Phase MovePhase 
SPLASH
Start Phase MessagePhase 
>> Feebas used Splash! 
Start Phase MoveEffectPhase 
Start Phase MessagePhase 
>> But nothing happened! 
Start Phase MoveEndPhase 
Start Phase MovePhase 
FAKE_OUT
Start Phase MessagePhase 
>> Wild Magikarp used Fake Out! 
Start Phase MoveEffectPhase 
base damage 32.239999999999995 Fake Out 40 45 50
damage 32 Fake Out
Start Phase DamageAnimPhase 
Start Phase MoveEndPhase 
Start Phase WeatherEffectPhase 
Start Phase BerryPhase 
Start Phase CheckStatusEffectPhase 
Start Phase TurnEndPhase 
```

</p>
</details>

Tempo-anon commented on test/abilities/steadfast.test.ts:
```diff
@@ -0,0 +1,87 @@
+
+    expect(playerPkm).not.toHaveFlinched();
+    expect(playerPkm).toHaveStatStage(Stat.SPD, 0);
+  });
```
Maybe another test for if the user cannot be flinched (like with a passive) would also be a good test

flx-sta replied:
Good point. I assume you mean [Inner Focus](https://bulbapedia.bulbagarden.net/wiki/Inner_Focus_(Ability))

flx-sta committed:
Update test/abilities/steadfast.test.ts

Co-authored-by: Tempoanon <163687446+Tempo-anon@users.noreply.github.com>

flx-sta committed:
Update test/abilities/steadfast.test.ts

Co-authored-by: Tempoanon <163687446+Tempo-anon@users.noreply.github.com>

flx-sta committed:
update Steadfast tests

- add test for "Inner Focus"
- add tests for "Inner Focus" being overriden by enemy ability

flx-sta committed:
update Steadfast tests

fix order of some expect checks to match tag check better

flx-sta committed:
Merge branch 'beta' into 721-abilitycheckup-steadfast

Tempo-anon commented on test/abilities/steadfast.test.ts:
```diff
@@ -0,0 +1,161 @@
+
+  it(`should NOT boost SPD if flinching is prevented by "Inner Focus"`, async () => {
+    const { classicMode, field, move, phaseInterceptor } = game;
+    game.override.passiveAbility(AbilityId.INNER_FOCUS);
```
nit: there's also shield dust, there's also the gen 9 item covert cloak (not implemented) that negates secondary effects

flx-sta replied:
Added Shield-Dust test and a todo test for Covert CLoak

Tempo-anon commented on test/abilities/steadfast.test.ts:
```diff
@@ -0,0 +1,161 @@
+
+    await phaseInterceptor.to("MoveEndPhase", true);
+
+    expect(playerPkm).toHaveFlinched();
```
Maybe out of scope for this PR, but does the flinching message still display if the flinched Pokemon has already acted?

flx-sta replied:
It does not. Showing the Test Output again:

## Test Output

<details><summary>[📄 Show Output]</summary>
<p>

```prolog
Start Phase EnemyCommandPhase 
Start Phase TurnStartPhase 
Start Phase MovePhase 
SPLASH
Start Phase MessagePhase 
>> Feebas used Splash! 
Start Phase MoveEffectPhase 
Start Phase MessagePhase 
>> But nothing happened! 
Start Phase MoveEndPhase 
Start Phase MovePhase 
FAKE_OUT
Start Phase MessagePhase 
>> Wild Magikarp used Fake Out! 
Start Phase MoveEffectPhase 
base damage 32.239999999999995 Fake Out 40 45 50
damage 32 Fake Out
Start Phase DamageAnimPhase 
Start Phase MoveEndPhase 
Start Phase WeatherEffectPhase 
Start Phase BerryPhase 
Start Phase CheckStatusEffectPhase 
Start Phase TurnEndPhase 
```

</p>
</details>

flx-sta committed:
add 2 Steadfast tests (1 todo)

- add coverage for "Shield Dust" ability
- add todo for Covert Cloak Item (not implemented yet)

PigeonBar commented on test/matchers/to-have-flinched.ts:
```diff
@@ -0,0 +1,33 @@
+    message: () =>
+      pass
+        ? `Expected ${received.name} to NOT have flinched, but it did!`
+        : `Expected ${received.name} to have flinched, but did not.`,
```
```suggestion
        : `Expected ${received.name} to have flinched, but it did not.`,
```

PigeonBar commented on test/abilities/steadfast.test.ts:
```diff
@@ -0,0 +1,174 @@
+
+    game.setTurnOrder([BattlerIndex.ENEMY, BattlerIndex.PLAYER]);
+    move.use(MoveId.SPLASH);
+    await move.selectEnemyMove(MoveId.SPLASH);
```
```suggestion
    await move.forceEnemyMove(MoveId.QUICK_ATTACK);
```

flx-sta replied:
May I ask why? I forced the order to be 
`    game.setTurnOrder([BattlerIndex.ENEMY, BattlerIndex.PLAYER]);` 
already

PigeonBar replied:
Just to make sure the test also works if the ability's owner took damage from an attack.

PigeonBar commented on test/matchers/to-have-stat-stage.ts:
```diff
@@ -0,0 +1,36 @@
+    pass,
+    message: () =>
+      pass
+        ? `Expected ${received.name} ${statStr} stage to NOT be ${statStr}, but it is!`
```
```suggestion
        ? `Expected ${received.name} ${statStr} stage to NOT be ${actualStage}, but it is!`
```

flx-sta replied:
Why, this is the negative/inverted case

PigeonBar replied:
So that it displays "Expected Pokemon's SPD to NOT be +1" instead of "Expected Pokemon's SPD to NOT be SPD".

flx-sta replied:
Oh. But it would probably be `${expectedStage}` then unless my brain is wired wrong right now
```suggestion
        ? `Expected ${received.name} ${statStr} stage to NOT be ${expectedStage}, but it is!`
```

PigeonBar replied:
That also works (`expectedStage === actualStage` when this test fails).

flx-sta replied:
Oh true. But from a logical perspective I’m locked on the expected value. Thanks for helping me resolve this

PigeonBar commented on src/data/battler-tags/flinched-tag.ts:
```diff
@@ -38,10 +36,12 @@ export class FlinchedTag extends BattlerTag {
         i18next.t("battlerTags:flinchedLapse", { pokemonNameWithAffix: getPokemonNameWithAffix(pokemon) }),
       );
+
+      applyAbAttrs<FlinchEffectAbAttr>(AbAttrFlag.FLINCH_EFFECT, pokemon, false);
```
```suggestion
      applyAbAttrs<FlinchEffectAbAttr>(AbAttrFlag.FLINCH_EFFECT, pokemon, false);

      return true;
```
<details><summary>According to Showdown, a Pokemon should be able to flinch twice in the same turn if it moves twice (e.g., due to Instruct)</summary>
<p>

![20250426 Double Flinch](https://github.com/user-attachments/assets/533d6fb5-6c31-4271-9658-36a58ce7195b)

</p>
</details>

flx-sta replied:
I see, I should probably have a test-case to cover that.
But how does `return true` fix that?

PigeonBar replied:
A tag gets removed whenever `lapse()` returns `false`, so returning `true` makes the tag remain after activating the flinch message.

flx-sta replied:
@PigeonBar Seems the way it works is correct, but the displayed messages are different.
Here is the ouput from a test of 
Magikarp + Magikarp
vs.
Feebas + Squirtle

```prolog
>> Wild Magikarp used Fake Out! 
Start Phase MoveEffectPhase 
base damage 34.85333333333333 Fake Out 40 44 45
damage 34 Fake Out
Start Phase DamageAnimPhase 
Start Phase MoveEndPhase 
Start Phase MovePhase 
SPLASH
Start Phase MessagePhase 
>> Wild Magikarp used Splash! 
Start Phase MoveEffectPhase 
Start Phase MessagePhase 
>> But nothing happened! 
Start Phase MoveEndPhase 
Start Phase MovePhase 
AERIAL_ACE
Start Phase MessagePhase 
>> Feebas flinched! 
Start Phase MoveEndPhase 
Start Phase MovePhase 
INSTRUCT
Start Phase MessagePhase 
>> Squirtle used Instruct! 
Start Phase MessagePhase 
>> But it failed! 
Start Phase MoveEndPhase 
Start Phase WeatherEffectPhase 
Start Phase BerryPhase 
Start Phase CheckStatusEffectPhase 
Start Phase TurnEndPhase 
Start Phase TurnInitPhase 
```

flx-sta committed:
Update test/matchers/to-have-flinched.ts

Co-authored-by: PigeonBar <56974298+PigeonBar@users.noreply.github.com>

flx-sta committed:
Merge branch 'beta' into 721-abilitycheckup-steadfast

flx-sta committed:
Update test/abilities/steadfast.test.ts

Co-authored-by: PigeonBar <56974298+PigeonBar@users.noreply.github.com>

flx-sta committed:
Update src/data/battler-tags/flinched-tag.ts

Co-authored-by: PigeonBar <56974298+PigeonBar@users.noreply.github.com>

flx-sta committed:
Merge branch 'beta' into 721-abilitycheckup-steadfast

flx-sta committed:
Update test/matchers/to-have-stat-stage.ts

flx-sta committed:
Merge branch 'beta' into 721-abilitycheckup-steadfast

flx-sta committed:
add flinch test

flx-sta committed:
Merge branch 'beta' into 721-abilitycheckup-steadfast

flx-sta committed:
chore: fix matchers.setup imports

flx-sta committed:
Merge branch 'beta' into 721-abilitycheckup-steadfast

flx-sta committed:
Merge branch 'beta' into 721-abilitycheckup-steadfast

flx-sta committed:
Merge branch 'beta' into 721-abilitycheckup-steadfast

PigeonBar commented on test/@types/vitest.d.ts:
```diff
@@ -85,5 +85,17 @@ declare module "vitest" {
     toHaveFullHp(): void;
+
+    /**
+     * Matcher to check if a {@linkcode Pokemon} has a sepcific {@linkcode Stats} stage.
```
```suggestion
     * Matcher to check if a {@linkcode Pokemon} has a specific {@linkcode Stats} stage.
```

PigeonBar commented on test/@types/vitest.d.ts:
```diff
@@ -85,5 +85,17 @@ declare module "vitest" {
+    toHaveStatStage(stat: Stat, expectedStage: number): void;
+
+    /**
+     * Matcher to check if a {@linkcode Pokemon} has flinched.
```
```suggestion
     * Matcher to check if a {@linkcode Pokemon} has flinched.
     *
     * (Specifically, this checks for the existence of a `FlinchedTag`. It does not check whether the flinch actually prevented the Pokemon from using a move.)
```

PigeonBar commented on test/flinch.test.ts:
```diff
@@ -0,0 +1,62 @@
+  it("should stay flinched if moving twice in a turn", async () => {
+    const { override, classicMode, move, phaseInterceptor } = game;
+
+    override.battleType("double").moveset([MoveId.SPLASH, MoveId.INSTRUCT, MoveId.AERIAL_ACE]);
```
```suggestion
    override.battleType("double").ability(AbilityId.DANCER);
```
It turns out that Instruct fails on Turn 1 because after Player 1 flinches, there is no valid move for it to repeat via Instruct. Instead, a Dancer setup will allow Player 1 to attempt to move twice on Turn 1.

```
>> Wild Magikarp used Fake Out!
Start Phase MoveEffectPhase
base damage 35.6 Fake Out 40 45 45
damage 35 Fake Out
Start Phase DamageAnimPhase
Start Phase MoveEndPhase
Start Phase MovePhase
SPLASH
Start Phase MessagePhase
>> Wild Magikarp used Splash!
Start Phase MoveEffectPhase
Start Phase MessagePhase
>> But nothing happened!
Start Phase MoveEndPhase
Start Phase MovePhase
SWORDS_DANCE
Start Phase MessagePhase
>> Feebas flinched!
Start Phase MoveEndPhase
Start Phase MovePhase
SWORDS_DANCE
Start Phase MessagePhase
>> Squirtle used Swords Dance!
Start Phase MoveEffectPhase
Start Phase StatStageChangePhase
Start Phase MessagePhase
>> Squirtle’s Attack sharply rose!
Start Phase ShowAbilityPhase
Start Phase MovePhase
SWORDS_DANCE
Start Phase MessagePhase
>> Feebas flinched!
Start Phase MoveEndPhase
Start Phase WeatherEffectPhase
Start Phase BerryPhase
Start Phase CheckStatusEffectPhase
Start Phase TurnEndPhase
Start Phase TurnInitPhase
Closing i18n MSW server!
```

PigeonBar commented on test/flinch.test.ts:
```diff
@@ -0,0 +1,62 @@
+
+    game.setTurnOrder([BattlerIndex.ENEMY, BattlerIndex.ENEMY_2, BattlerIndex.PLAYER, BattlerIndex.PLAYER_2]);
+    move.select(MoveId.AERIAL_ACE, BattlerIndex.PLAYER);
+    move.select(MoveId.INSTRUCT, BattlerIndex.PLAYER_2, BattlerIndex.PLAYER);
```
```suggestion
    move.use(MoveId.SWORDS_DANCE, BattlerIndex.PLAYER);
    move.use(MoveId.SWORDS_DANCE, BattlerIndex.PLAYER_2);
```

PigeonBar commented on test/flinch.test.ts:
```diff
@@ -0,0 +1,62 @@
+    expect(player1).toHaveFlinched();
+    await game.toEndOfTurn();
+
+    expect(player1).not.toHaveFlinched(); // tag was lapsed
```
```suggestion
    expect(player1).not.toHaveFlinched(); // tag was lapsed

    // Check that Player 1 attempted to copy Swords Dance but could not move due to flinching
    expect(player1.summonData.abilitiesApplied).toContain(AbilityId.DANCER);
    expect(player1).toHaveStatStage(Stat.ATK, 0);
```

PigeonBar commented on test/flinch.test.ts:
```diff
@@ -0,0 +1,62 @@
+import { BattlerIndex } from "#enums/battler-index";
+import { MoveId } from "#enums/move-id";
+import { SpeciesId } from "#enums/species-id";
```
```suggestion
import { AbilityId } from "#enums/ability-id";
import { BattlerIndex } from "#enums/battler-index";
import { MoveId } from "#enums/move-id";
import { SpeciesId } from "#enums/species-id";
import { Stat } from "#enums/stat";
```

flx-sta committed:
Merge branch 'beta' into 721-abilitycheckup-steadfast

innerthunder commented on src/data/abilities/ab-attrs/flinch-stat-stage-change-ab-attr.ts:
```diff
@@ -4,8 +4,16 @@ import { StatStageChangePhase } from "#app/phases/stat-stage-change-phase";
+ * @see {@link https://bulbapedia.bulbagarden.net/wiki/Steadfast_(Ability) Steadfast - Bulbapedia}
+ */
 export class FlinchStatStageChangeAbAttr extends FlinchEffectAbAttr {
+  /** The {@linkcode BattleStat stats} to change. */
```
```suggestion
  /** The {@linkcode BattleStat | stats} to change. */
```

innerthunder commented on src/data/abilities/ab-attrs/flinch-stat-stage-change-ab-attr.ts:
```diff
@@ -4,8 +4,16 @@ import { StatStageChangePhase } from "#app/phases/stat-stage-change-phase";
+  /** The {@linkcode BattleStat stats} to change. */
   private readonly stats: BattleStat[];
+
+  /** The number of stages to apply to the {@linkcode stats}. */
```
Not sure if the link works here

flx-sta replied:
It does, in vsc

innerthunder commented on test/@types/vitest.d.ts:
```diff
@@ -92,5 +92,17 @@ declare module "vitest" {
+    /**
+     * Matcher to check if a {@linkcode Pokemon} has a sepcific {@linkcode Stats} stage.
+     * @param stat - The {@linkcode Stat} to check
+     * @param expectedStage - The expected stage of the {@linkcode stat}
```
Not sure if the "stat" link works here

flx-sta replied:
It does. It simply links to the param inside the method. No real gain, but also no loss

innerthunder commented on test/test-utils/matchers/to-have-flinched.ts:
```diff
@@ -0,0 +1,33 @@
+ * @param received - The object to check. Should be a {@linkcode Pokemon}.
+ * @returns Whether the matcher passed
+ */
+export function toHaveFlinchedMatcher(this: MatcherState, received: unknown): SyncExpectationResult {
```
No harm in this, but imo we should be using a more generic `ToHave[Battler]Tag` matcher instead.

flx-sta replied:
Agreed. This would also match the additional description for the use that was given by PigeonBar

flx-sta committed:
Merge branch 'beta' into 721-abilitycheckup-steadfast

flx-sta committed:
Update src/data/abilities/ab-attrs/flinch-stat-stage-change-ab-attr.ts

Co-authored-by: innerthunder <168692175+innerthunder@users.noreply.github.com>

flx-sta committed:
rename `toHaveFlinched` to `toHaveBattlerTagType`

flx-sta committed:
add `toHaveAbilityApplied` matcher

flx-sta committed:
update move-phase console.log

now includes the pokemon name with affix

flx-sta committed:
Update matchers output

flx-sta committed:
Add undefined check to toHaveUsedMove matcher

flx-sta committed:
Merge branch 'beta' into 721-abilitycheckup-steadfast

flx-sta committed:
Merge branch 'beta' into 721-abilitycheckup-steadfast

flx-sta committed:
Merge branch 'beta' into 721-abilitycheckup-steadfast
